### PR TITLE
added picca_nersc_submit.py script

### PR DESCRIPTION
**Pull request by ngbusca**
This branch adds the picca_nersc_submit script:
 * it creates the output directory and all the necessary .batch files all the way to the fitter (basic fit without shenanigans)
 * it creates a submission script with the right dependencies (e.g. do_cf.py waits for do_deltas.py, fitter2 waits for do_cf, do_dmat, and export, etc.)
 * optional `--debug` option for a short run over 10,000 quasars
 * for extra safety, the submission is not automatic (the user has to run the created submit.sh script)
 * keeps you informed of the progress by email (optional)
 * choice of doing everything, only `cf` or only `xcf` analysis
 * the user can choose the redshift intervals for the analysis (the default is four redshift bins)
 * the current weak point is the running times which are hardcoded to the default four bins. It will be ok with more redshfit bins but it might fail with less. 
 * I'm still working on a QA python notebook

Example run:

    picca_nersc_submit --in-dir <some_path>/spectra-16 --drq <some_path>/drq.fits --out-dir <outdir> --email <your@email>`
    cd <outdir>
    ./submit.sh

### CODE REVIEW

**Comment by andreufont** [comment]
I added a related issue in #487

**Comment by julienguy** [comment]
Some comments on my tests:
 - need nargs="*" for --zint option in picca_nersc_submit
 - I changed the scripts to run python with -u option, "python -u `which do_deltas.py`" for instance, in order to have logging messages in real time to see what is going on.
 - fitter2 ini file has hardcoded path to model ```/global/homes/n/nbusca/igmhub/picca/py/picca/fitter2/models/PlanckDR12/PlanckDR12.fits``` ( should think about giving only a relative path in config file, like ``PlanckDR12/PlanckDR12.fits```, and then use 
```
from pkg_resources import resource_filename
filename=os.path.join(resource_filename('picca',"/fitter2/models"),relative_path)
```
- I haven't been run the correlation function in 4 hours (while running with a single bin --zint 0:10)
I am trying a 10h job. We should resume the work on the mpi version of do_cf.py.

**Comment by ngbusca** [comment]
@julienguy

>> need nargs="*" for --zint option in picca_nersc_submit

Fixed

>> I changed the scripts to run python with -u option, "python -u which do_deltas.py" for instance, in order to have logging messages in real time to see what is going on.

I'll look into that

>> fitter2 ini file has hardcoded path to model /global/homes/n/nbusca/igmhub/picca/py/picca/fitter2/models/PlanckDR12/PlanckDR12.fits ( should think about giving only a relative path in config file, like ``PlanckDR12/PlanckDR12.fits```, and then use
from pkg_resources import resource_filename
filename=os.path.join(resource_filename('picca',"/fitter2/models"),relative_path)

Good idea, I'll fix that

>> I haven't been run the correlation function in 4 hours (while running with a single bin --zint 0:10)
I am trying a 10h job. We should resume the work on the mpi version of do_cf.py.

Or you can run inseveral z bins and coadd them (it's strictly equivalent) : a poor man's mpi...

**Comment by ngbusca** [comment]
I'll work on a coaddition code for several z bins (I think @londumas already has a version of that)

**Comment by julienguy** [comment]
That would be great. I was just in the process of running Helion's export_stacked_correlation.py on your binned correlation functions of L4.

**Comment by londumas** [comment]
@ngbusca and @julienguy, fitter2 by default goes to find the power-spectrum inside picca: `resource_filename('picca', 'fitter2/models/PlanckDR12/PlanckDR12.fits')`.
If someones runs it it will use its own local power-spectrum, not the one of @ngbusca.

File: bin/picca_nersc_submit [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+        fout.write("croom_par1 = 0.289 0.1 None None fixed\n")
+    fout.close()
+
+    chi2_ini = outdir+"/chi2_{}".format(cf_file.replace(".fits",".ini"))
+    fout = open(chi2_ini,"w")
+    fout.write("[data sets]\n")
+    fout.write("zeff = 2.310\n")
+    fout.write("ini files = {}\n".format(cf_file.replace(".fits",".ini")))
+
+    fout.write("[fiducial]\n")
```
Comment (londumas) [inline bin/picca_nersc_submit]: You can remove these two lines. picca will take the local power spectrum by default

**Comment by londumas** [comment]
@ngbusca, for some reason I don't understand, I can't get the cell with the table of best fit running using `xcf` when using the jupyter notebook. However, it works running the lines on ipython.
The issue is with `xcf_z_0_2.65`. Maybe you could try rerunning the fit.
```
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-21-c571de57d42b> in <module>()
     15 print("{:20}".format("zeff"),end="")
     16 for f in fi:
---> 17     ff = h5py.File(f,'r')
     18     zeff = ff['best fit'].attrs['zeff']
     19     out = "{}".format(round(zeff,2))

~/.conda/envs/picca_plots/lib/python3.6/site-packages/h5py/_hl/files.py in __init__(self, name, mode, driver, libver, userblock_size, swmr, **kwds)
    267             with phil:
    268                 fapl = make_fapl(driver, libver, **kwds)
--> 269                 fid = make_fid(name, mode, userblock_size, fapl, swmr=swmr)
    270 
    271                 if swmr_support:

~/.conda/envs/picca_plots/lib/python3.6/site-packages/h5py/_hl/files.py in make_fid(name, mode, userblock_size, fapl, fcpl, swmr)
     97         if swmr and swmr_support:
     98             flags |= h5f.ACC_SWMR_READ
---> 99         fid = h5f.open(name, flags, fapl=fapl)
    100     elif mode == 'r+':
    101         fid = h5f.open(name, h5f.ACC_RDWR, fapl=fapl)

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/h5f.pyx in h5py.h5f.open()

OSError: Unable to open file (unable to lock file, errno = 11, error message = 'Resource temporarily unavailable')
```

**Comment by ngbusca** [comment]
@londumas this means that at some point you opened it without the 'r' option and it's now locked or probably dammaged. I noticed the issue (see the warning on top of the notebook) I'm looking into how to permanently prevent it.

**Comment by londumas** [comment]
@ngbusca, ok. I think I corrected all of the previous open without 'r'.
It is strange however that I can run on the terminal but not on the notebook.

**Comment by ngbusca** [comment]
@londumas let me know whenI can take a look

**Comment by londumas** [comment]
@ngbusca, you can.

**Comment by ngbusca** [comment]
@londumas I did some cleaning on the notebook. Please take a look and tell me if you think the notebook is ready for merging to master (there are still some improvements needed on the picca_nersc_submit). If you make any changes, remember to clear clear all the output before committing.

**Comment by londumas** [comment]
@ngbusca, I just looked at your commit and it seems perfect.
I just added looking at the covariance and correlation matrix.
Ready to be merged if ok with you.

**Comment by ngbusca** [comment]
@londumas `export_stacked_correlation.py` script doesn't coadd the distortion matrices. Is it ok if I modify that? Would it be ok to modify it such that:
 * If only one dmat file is given, then keep the current behavior
 * If as many dmats as data are given then coadd the dmats assuming that they are given in the same order as the data
 * If no dmats are given, try to "discover" them from the data filename (replacing cf by dmat in the name). If some dmats but not all are found, error and exit. If no dmats are found, use the identity

**Comment by ngbusca** [comment]
@julienguy as a more permanent solution to flushing stdout, I've overridden the print function such that it's flushed. I've replaced all the zoo of print, sys.stdout.write and sys.stderr.write with the single picca.utils.print function. For future code we should stick to using the picca.utils.print function.

**Comment by londumas** [comment]
@ngbusca, of course do as you please with `export_stacked_correlation.py`

**Comment by londumas** [comment]
@ngbusca, I don't understand why you add a new function  `bin/coadd_zint`. Why not simply move `export_stacked_correlation.py` to `bin/coadd_zint`?
My function also takes care of the case where all function don't have the same number of healpix pixels.

**Comment by ngbusca** [comment]
@londumas yes, your code is very complete! I was just trying to set something working before merging  with yours.

**Comment by ngbusca** [comment]
I will have very little time to work on this PR in the coming weeks, and I would like to get it merged to master. @londumas we could add what you request in a future PR, specially since stacking and computing the cf1d take seconds and can be done on the command line.

By the way, the current `bin/coadd_zint` now takes correctly into account the potential different healpixelizations in the input files.

If there are no objections, can @londumas or @andreufont approve so I can merge?

**Comment by andreufont** [comment]
Yes, we can accept this pull request and open new issues for these extra features.

**Pull request merged**